tests/test_observable/test_catch.py

```diff
@@ -14,112 +15,119 @@
 
 class TestCatch(unittest.TestCase):
 
-    def test_catch_no_errors(self):
+    def _base_catch_exception_test(self, msgs, expected_msgs=None, use_chaining_api=True):
+        """Given lists of messages for successive observables, creates the
+        appropriate observables and chains them together using catch_exception.
+        Tests that the actions seen by a subscriber match the expected
+        messages.
+
+        Params:
+
+        :param msgs:
+            List of lists of expected messages.
+        :param expected_msgs:
+            List of expected messages received by subscribers. If not provided,
+            a "best guess" is used. Note that this "best guess" currently
+            doesn't perform well with non-terminating Observables (i.e.
+            Observable.never()).
+        :param use_chaining_api:
+            Whether to construct the catch_exception call by repeated chaining
+            off of instance methods, or by calling Observable.catch_exception
+            at the class level.
+        """
         scheduler = TestScheduler()
-        msgs1 = [on_next(150, 1), on_next(210, 2), on_next(220, 3), on_completed(230)]
-        msgs2 = [on_next(240, 5), on_completed(250)]
-        o1 = scheduler.create_hot_observable(msgs1)
-        o2 = scheduler.create_hot_observable(msgs2)
+
+        def _ms_dispatch(ms):
+            if ms == []:
+                return Observable.never()
+            else:
+                return scheduler.create_hot_observable(ms)
+
+        os = [_ms_dispatch(ms) for ms in msgs]
 
         def create():
-            return o1.catch_exception(o2)
+            if use_chaining_api:
+                return reduce(lambda o, catcher: o.catch_exception(catcher), os)
```

Comment:
RxPy is compatible with Python 2.7 and 3.4 so its ok to import reduce from functools.

Reply:
done.